node/rpc/src/state_api.rs

```diff
@@ -186,3 +190,91 @@ pub(crate) async fn state_wait_msg<DB: Blockstore + Clone + Send + Sync + 'stati
         return_dec: IpldJson(ipld),
     })
 }
+
+// Sample CIDs (useful for testing):
+//   Mainnet:
+//     1,594,681 bafy2bzaceaclaz3jvmbjg3piazaq5dcesoyv26cdpoozlkzdiwnsvdvm2qoqm OhSnap upgrade
+//     1_960_320 bafy2bzacec43okhmihmnwmgqspyrkuivqtxv75rpymsdbulq6lgsdq2vkwkcg Skyr upgrade
+//     2,833,266 bafy2bzacecaydufxqo5vtouuysmg3tqik6onyuezm6lyviycriohgfnzfslm2
+//     2,933,266 bafy2bzacebyp6cmbshtzzuogzk7icf24pt6s5veyq5zkkqbn3sbbvswtptuuu
+//   Calibnet:
+//     242,150 bafy2bzaceb522vvt3wo7xhleo2dvb7wb7pyydmzlahc4aqd7lmvg3afreejiw
+//     630,932 bafy2bzacedidwdsd7ds73t3z76hcjfsaisoxrangkxsqlzih67ulqgtxnypqk
+//
+/// Traverse an IPLD directed acyclic graph and use libp2p-bitswap to request any missing nodes.
+/// This function has two primary uses: (1) Downloading specific state-roots when Forest deviates
+/// from the mainline blockchain, (2) fetching historical state-trees to verify past versions of the
+/// consensus rules.
+pub(crate) async fn state_fetch_root<DB: Blockstore + Clone + Sync + Send + 'static, B: Beacon>(
+    data: Data<RPCState<DB, B>>,
+    Params((CidJson(root_cid),)): Params<StateFetchRootParams>,
+) -> Result<StateFetchRootResult, JsonRpcError> {
+    const MAX_CONCURRENT_REQUESTS: usize = 16;
+    const REQUEST_TIMEOUT: Duration = Duration::from_secs(10);
+
+    let sem = Arc::new(Semaphore::new(MAX_CONCURRENT_REQUESTS));
+    let mut seen: HashSet<Cid> = HashSet::new();
+    let mut counter: usize = 0;
+    let mut failures: usize = 0;
+    let mut task_set = JoinSet::new();
+
+    let mut get_ipld_link = |ipld: &Ipld| match ipld {
+        Ipld::Link(cid) if cid.codec() == 0x71 && seen.insert(*cid) => Some(*cid),
+        _ => None,
+    };
+
+    task_set.spawn(async move { Ok(Ipld::Link(root_cid)) });
+
+    // Iterate until there are no more ipld nodes to traverse
+    while let Some(result) = task_set.join_next().await {
+        match result? {
+            Ok(ipld) => {
+                for new_cid in ipld.iter().filter_map(&mut get_ipld_link) {
+                    counter += 1;
+                    if counter % 1_000 == 0 {
+                        // set RUST_LOG=forest_rpc::state_api=debug to enable these printouts.
```

Comment:
Maybe leave a TODO and refer to https://github.com/ChainSafe/forest/issues/2955 if it's meant to be replaced by the desired progress bar

Reply:
Showing a progress bar requires knowing how much work is left. We should show some progress indicators in the client, though. Maybe as a separate PR.